cloudinit/sources/DataSourceAzure.py

```diff
@@ -208,6 +207,33 @@ def get_hv_netvsc_macs_normalized() -> List[str]:
     ]
 
 
+@azure_ds_telemetry_reporter
+def determine_device_driver_for_mac(mac: str) -> Optional[str]:
+    """Determine the device driver to match on, if any."""
+    drivers = [
+        i[2]
+        for i in net.get_interfaces(blacklist_drivers=BLACKLIST_DRIVERS)
```

Comment:
Is there some way we can hang this off of DataSourceAzure and call `self.distro.networking.get_interfaces_by_mac()` here which should already account for blacklist_drivers? We are trying to move general networking operations into the `distro.networking` as the primary entry point as different distros will have unique APIs or filesystems to get at that device content. Eventually I think we want to retire cloudinit.net functions in favor of distro.networking driven API.

Reply:
A note as well, we will be modernizing our variable names in the future to be something more inclusive and blacklist will likely be replaced with `denylist` throughout cloud-init in the future. so expect a couple of merge conflicts if that PR lands before yours.

Reply:
Will fix & revisit after 22.3

Reply:
> Is there some way we can hang this off of DataSourceAzure and call `self.distro.networking.get_interfaces_by_mac()` here which should already account for blacklist_drivers? We are trying to move general networking operations into the `distro.networking` as the primary entry point as different distros will have unique APIs or filesystems to get at that device content. Eventually I think we want to retire cloudinit.net functions in favor of distro.networking driven API.

I looked at get_interfaces_by_mac, but it doesn't include the driver afaict.

Reply:
I meant [cloudinit.distros.networking.Networking.get_interfaes_by_mac](https://github.com/canonical/cloud-init/blob/main/cloudinit/distros/networking.py#L74-L78) which includes [Azure blacklist_drivers settings](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceAzure.py#L712)

Reply:
> I meant [cloudinit.distros.networking.Networking.get_interfaes_by_mac](https://github.com/canonical/cloud-init/blob/main/cloudinit/distros/networking.py#L74-L78) which observes [Azure blacklist_drivers settings](https://github.com/canonical/cloud-init/blob/main/cloudinit/sources/DataSourceAzure.py#L712)

Yeah, it filters, but I need the actual driver for the interface too, not just the name (which is all that is contained within the dict).  If we can get it to return that namedtuple or whatever you were thinking of that would include the driver name, that'd work perfectly.

Reply:
@cjp256 thanks I had misinterpreted the output format from get_interfaces_by_mac. You are correct in this PR, there is no better place at the moment than to call the `net.get_interfaces()`. I was going to pick at the repeated calls to `net.get_interfaces()` here per device representing in IMDS as "costly" to walk /sys/class/net devices on bigger systems. But, from a wall clock time cost, it seems to be ~4 milliseconds per call. so it's not going to be noticeable in the grand scheme of things. Note, this call-site may migrate into the distro.networking classes in the future just like [cloudinit.distros.networking.Networking.get_interfaces_by_mac](https://github.com/canonical/cloud-init/blob/main/cloudinit/distros/networking.py#L74-L78)  per the [network/distro refactor work](https://github.com/canonical/cloud-init/blob/main/WIP-ONGOING-REFACTORIZATION.rst).